src/dune_rules/gen_rules.ml

```diff
@@ -375,22 +375,25 @@ let gen_rules ~sctx ~dir components : Build_system.extra_sub_directories_to_keep
     [ subdirs_to_keep1; subdirs_to_keep2; subdirs_to_keep3 ]
 
 let filter_out_stanzas_from_hidden_packages ~visible_pkgs =
-  List.filter_map ~f:(fun stanza ->
+  List.concat_map ~f:(fun stanza ->
       match Dune_file.stanza_package stanza with
-      | None -> Some stanza
-      | Some package -> (
-        if Package.Name.Map.mem visible_pkgs package.name then
-          Some stanza
-        else
+      | None -> [ stanza ]
+      | Some package ->
+        let redirect =
           match stanza with
-          | Library l ->
+          | Library l -> (
             (* A public library should still be referable by its private name
                even we filter it out. Therefore, we create a private -> public
                name mapping for the filtered libraries *)
-            let open Option.O in
-            let+ dln = Library_redirect.Local.of_lib l in
-            Library_redirect dln
-          | _ -> None ))
+            match Library_redirect.Local.of_lib l with
+            | None -> []
+            | Some dln -> [ Library_redirect dln ] )
+          | _ -> []
```

Comment:
Yes, you are right. To be honest, I have no idea what I was thinking here since the entire point of the PR is to move all the redirections to one place.